core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -197,19 +204,25 @@ class KafkaApis(val requestChannel: RequestChannel,
     val stopReplicaRequest = request.body[StopReplicaRequest]
 
     if (isAuthorizedClusterAction(request)) {
-      val (result, error) = replicaManager.stopReplicas(stopReplicaRequest)
-      // Clearing out the cache for groups that belong to an offsets topic partition for which this broker was the leader,
-      // since this broker is no longer a replica for that offsets topic partition.
-      // This is required to handle the following scenario :
-      // Consider old replicas : {[1,2,3], Leader = 1} is reassigned to new replicas : {[2,3,4], Leader = 2}, broker 1 does not receive a LeaderAndIsr
-      // request to become a follower due to which cache for groups that belong to an offsets topic partition for which broker 1 was the leader,
-      // is not cleared.
-      result.foreach { case (topicPartition, error) =>
-        if (error == Errors.NONE && stopReplicaRequest.deletePartitions && topicPartition.topic == GROUP_METADATA_TOPIC_NAME) {
-          groupCoordinator.handleGroupEmigration(topicPartition.partition)
+      if (controller.isCurrentOrUnknownBrokerEpoch(stopReplicaRequest.brokerEpoch())) {
```

Comment:
The code can probably be more readable with less nested if/else by doing this:

```
if (!isAuthorizedClusterAction(request)) {
  ...
} else if (!controller.isCurrentOrUnknownBrokerEpoch(...)) {
  ...
} else {
  ...
}
```

Same for other methods.

Reply:
That is a good point. Thanks for the suggestion. Done.